runtime/src/main/java/org/corfudb/runtime/view/TableRegistry.java

```diff
@@ -0,0 +1,170 @@
+import org.corfudb.runtime.collections.CorfuTable;
+import org.corfudb.runtime.collections.Table;
+import org.corfudb.runtime.collections.TableOptions;
+import org.corfudb.runtime.object.transactions.TransactionType;
+import org.corfudb.util.serializer.ISerializer;
+import org.corfudb.util.serializer.ProtobufSerializer;
+import org.corfudb.util.serializer.Serializers;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * Created by zlokhandwala on 2019-08-10.
+ */
+public class TableRegistry {
+
+    private static final String CORFU_SYSTEM_NAMESPACE = "CorfuSystem";
+    private static final String REGISTRY_TABLE_NAME = "RegistryTable";
+
+    private final CorfuRuntime runtime;
+
+    private final Map<String, Class<? extends Message>> classMap;
+
+    private final Map<String, Table<Message, Message, Message>> tableMap;
+
+    private final ISerializer protobufSerializer;
+
+    private final CorfuTable<TableName, CorfuRecord<TableDescriptors, Message>> registryTable;
+
+    public TableRegistry(CorfuRuntime runtime) {
+        this.runtime = runtime;
+        this.classMap = new ConcurrentHashMap<>();
+        this.tableMap = new ConcurrentHashMap<>();
+        this.protobufSerializer = new ProtobufSerializer((byte) 25, classMap);
+        Serializers.registerSerializer(this.protobufSerializer);
+        this.registryTable = this.runtime.getObjectsView().build()
+                .setTypeToken(new TypeToken<CorfuTable<TableName, CorfuRecord<TableDescriptors, Message>>>() {
+                })
+                .setStreamName(getFullyQualifiedTableName(CORFU_SYSTEM_NAMESPACE, REGISTRY_TABLE_NAME))
+                .setSerializer(this.protobufSerializer)
+                .open();
+
+        // Register the table schemas to schema table.
+        addTypeToClassMap(TableName.getDefaultInstance());
+        addTypeToClassMap(TableDescriptors.getDefaultInstance());
+
+        registerTable(CORFU_SYSTEM_NAMESPACE,
+                REGISTRY_TABLE_NAME,
+                TableName.getDescriptor().toProto(),
+                TableDescriptors.getDescriptor().toProto());
+    }
+
+    private void registerTable(String namespace,
+                               String tableName,
+                               DescriptorProto keyDescriptor,
+                               DescriptorProto valueDescriptor) {
+
+        TableName tableNameKey = TableName.newBuilder()
+                .setNamespace(namespace)
+                .setTableName(tableName)
+                .build();
+        TableDescriptors tableDescriptors = TableDescriptors.newBuilder()
+                .setKeyDescriptor(keyDescriptor)
+                .setValueDescriptor(valueDescriptor)
+                .build();
+
+        try {
+            this.runtime.getObjectsView().TXBuild().type(TransactionType.OPTIMISTIC).build().begin();
+            this.registryTable.putIfAbsent(tableNameKey,
+                    new CorfuRecord<>(tableDescriptors, null));
```

Comment:
is it possible to avoid using null?

Reply:
this is from client's api, if not null then we will force them to pass either an Optional or a dummy metadata, both of which can be more complex than null which they have asked for right now.